File: Phase-0/reservieren/reservieren/Models/Book.cs
namespace reservieren.Models
{
    using Akka.Actor;
    using Akka.Persistence;

    public partial class Book : UntypedPersistentActor
    {
        private string Id;

        private string Isbn;

        private string Title;

        private string Author;

        private string ShortDescription;

        private bool Lent;

        private bool Reserved;

        public static Props props => Props.Create(() => new Book());

        //protected ILoggingAdapter Log { get; } = Context.GetLogger();

        public override string PersistenceId { get; } = Context.Self.Path.Name;

        protected override void PreStart()
        {
            base.PreStart();
            Log.Info($"Book: {Context.Self.Path.Name}");
        }

        protected override void OnCommand(object message)
        {
            switch (message)
            {
                case BookData book:

                    if (Id != book.Id)
                    {
                        Persist(new BookIdUpdated(book.Id), evt => Id = book.Id);
                    }

                    if (Isbn != book.Isbn)
                    {
                        Persist(new BookIsbnUpdated(book.Isbn), evt => Isbn = book.Isbn);
                    }

                    if (Title != book.Titel)
                    {
                        Persist(new BookTitleUpdated(book.Titel), evt => Title = book.Titel);
                    }

                    if (Author != book.Autor)
                    {
                        Persist(new BookAuthorUpdated(book.Autor), evt => Author = book.Autor);
                    }

                    if (ShortDescription != book.KurzBeschreibung)
                    {
                        Persist(new BookShortDescriptionUpdated(book.KurzBeschreibung),
                                evt => ShortDescription = book.KurzBeschreibung);
                    }
                    break;
                case BookLent bookLent:
                    if (Lent != bookLent.Lent)
                    {
                        Persist(new BookLentUpdated(bookLent.Lent), evt => Lent = bookLent.Lent);
                    }
                    if (Isbn != bookLent.Isbn)
                    {
                        Persist(new BookIsbnUpdated(bookLent.Isbn), evt => Isbn = bookLent.Isbn);
                    }
                    break;
                case Reserve _:
                    if (!Reserved)
                    {
                        Persist(new BookReservedUpdated(true), evt => Reserved = evt.Reserved);
                    }
                    Sender.Tell(new BookReserved(true));
                    break;
                case CancelReserve _:

                    if (Reserved)
                    {
                        Persist(new BookReservedUpdated(false), evt => Reserved = evt.Reserved);
                    }
                    Sender.Tell(new CanceledReserve(true));
                    break;
            }
        }

        protected override void OnRecover(object message)
        {
            switch (message)
            {
                case BookIdUpdated evt:
                    Id = evt.Id;
                    break;
                case BookIsbnUpdated evt:
                    Isbn = evt.Isbn;
                    break;
                case BookTitleUpdated evt:
                    Title = evt.Title;
                    break;
                case BookAuthorUpdated evt:
                    Author = evt.Author;
                    break;
                case BookShortDescriptionUpdated evt:
                    ShortDescription = evt.ShortDescription;
                    break;
                case BookLentUpdated evt:
                    Lent = evt.Lent;
                    break;
                case BookReservedUpdated evt:
                    Reserved = evt.Reserved;
                    break;
            }
        }
    }
}

File: Phase-0/ausleihen/ausleihen-scs/src/main/scala/container/bootcamp/ausleihen/books/Book.scala
package container.bootcamp.ausleihen.books

import akka.actor.{ActorLogging, Props}
import akka.persistence.PersistentActor
import container.bootcamp.ausleihen.books.Book.BookEvents._
import container.bootcamp.ausleihen.books.Book._


/**
  * Book actor which holds a state for a dedicated book only the update stuff.
  */
object Book {
  def props: Props = Props(new Book)
  case class BookData(
                       id: Option[String],
                       Isbn: String,
                       Titel: String,
                       Autor: String,
                       KurzBeschreibung: String,
                       lent: Boolean = false,
                       reserved: Boolean = false
                     )
  case object BookLend
  case class BookLent(lend: Boolean)
  case object BookAlreadyLent
  case object BookReturn
  case class BookReturned(returned: Boolean)
  case object BookAlreadyReturned
  case object BookDataGet
  case class BookReserved(Reserved: Boolean, Isbn: String)

  /*
   * Events which are persisted for a book
   */
  object BookEvents {
    sealed abstract class BookEvent
    case class BookIdUpdated(id: String) extends BookEvent
    case class BookIsbnUpdated(isbn: String) extends BookEvent
    case class BookTitleUpdated(title: String) extends BookEvent
    case class BookAuthorUpdated(author: String) extends BookEvent
    case class BookDescriptionUpdated(description: String) extends BookEvent
    case class BookLentUpdated(lend: Boolean) extends BookEvent
    case class BookReservedUpdated(reserved: Boolean) extends BookEvent

  }

}
class Book extends PersistentActor with ActorLogging {

  override def persistenceId: String = self.path.name

  var id: Option[String] = None
  var isbn: String = _
  var title: String = _
  var author: String = _
  var description: String = _
  var lent: Boolean = _
  var reserved: Boolean = _

  def receiveCommand: PartialFunction[Any, Unit] = {
    case BookData(bdId, bdIsbn, bdTitle, bdAuthor, bdDescription, bdLend, bdReserved) =>

      bdId.foreach { id =>
        if(this.id != bdId) persist(BookIdUpdated(id)) { _ => this.id = bdId }
      }

      if(isbn != bdIsbn) persist(BookIsbnUpdated(bdIsbn)) {e => isbn = e.isbn}
      if(title != bdTitle) persist(BookTitleUpdated(bdTitle)) {e => title = e.title}
      if(author != bdAuthor) persist(BookAuthorUpdated(bdAuthor)) {e => author = e.author}
      if(description != bdDescription) persist(BookDescriptionUpdated(bdDescription)) {e => description = e.description}
      if(lent != bdLend) persist(BookLentUpdated(bdLend)) { e => lent = e.lend}
      if(reserved != bdReserved) persist(BookReservedUpdated(bdReserved)){e => reserved = e.reserved}

    case BookLend =>

      if(!lent) {
        persist(BookLentUpdated(true)) { e =>
            lent = e.lend
            sender ! BookLent(lent)
        }
      } else {
        sender ! BookAlreadyLent
      }

    case BookReturn =>

      if(lent) {
        persist(BookLentUpdated(false)) { e =>
            lent = e.lend
            sender ! BookReturned(lent)
        }
      } else {
        sender ! BookAlreadyReturned
      }
    case BookReserved(bookReserved,bookIsbn) =>
      persist(BookReservedUpdated(bookReserved)) { e =>
        reserved = e.reserved
        if(isbn != bookIsbn) {
          if(isbn != bookIsbn) persist(BookIsbnUpdated(bookIsbn)) {e => isbn = e.isbn}
        }

      }

  }

  def receiveRecover: PartialFunction[Any, Unit] = {
    case BookIdUpdated(bdId) => id = Some(bdId)
    case BookIsbnUpdated(bdIsbn) => isbn = bdIsbn
    case BookTitleUpdated(bdTitle) => title = bdTitle
    case BookAuthorUpdated(bdAuthor) => author = bdAuthor
    case BookDescriptionUpdated(bdDescription) => description = bdDescription
    case BookLentUpdated(bdLend) => lent = bdLend
    case BookReservedUpdated(bdReserved) => reserved = bdReserved
  }
}
